### PR TITLE
[otbn] Specify reaction to memory accesses while OTBN is busy

### LINKED ISSUE #2696: [OTBN/SW] Cause a bus error on access to OTBN memories whilst OTBN is active
The initial OTBN implementation blocks any access to the OTBN I and D mem whilst the OTBN core is busy. This seems like a sensible design but has the side-effect that if Ibex attempts to access either of these memories whilst the OTBN core is busy it will stall until OTBN finishes.

I propose any access to the OTBN memories whilst it is running should result in a bus error. It seems any attempt by software to access the memories whilst it is running would be a bug and it is preferable that results in an exception rather than Ibex stalling for an indeterminate period. If there's a bug in the OTBN software then it may never terminate and the whole system would lock up.

> estimate 4
> remaining 2021-08-09 2

### PR DESCRIPTION
**Pull request by imphil**
OTBN has a data and an instruction memory, both of which are meant to be
"inaccessible" as long as OTBN is executing a command. The question is:
what happens if host software still tries to access the memory?

Currently, we vaguely specify that we'll "block" the access, i.e. hang
the bus.

In further discussions we agreed to instead issue a fatal error, which
follows the normal fatal error machinery: terminate the current
operation (if any), indicate the problem in the FATAL_ALERT_CAUSE
register, and sometimes in the ERR_BITS register.

Fixes #2696

### CODE REVIEW

**Comment by imphil** [comment]
This is the spec change only for now to get initial feedback. If we're happy with it it probably makes sense to add the RTL to this PR as well to keep things in sync.

**Review by GregAC — APPROVED** [review]
Related to this is the behaviour when OTBN reads from the scrambled dmem/imem and doesn't get a grant. This should only happen if a key isn't setup and I believe we'd agreed this would also cause OTBN to halt and signal an error (recoverable rather than fatal alert).

File: hw/ip/otbn/data/otbn.hjson [hunk 1]
```diff
@@ -234,6 +234,14 @@
           name: "fatal_reg"
           desc: "A fatal error was seen on a GPR or WDR read."
         }
+        { bits: "8",
+          name: "fatal_illegal_bus_access"
+          resval: 0,
+          desc: '''
+            A bus-accessible register was read or written while when it was not
```
Comment (rswarbrick) [inline hw/ip/otbn/data/otbn.hjson:241]: Register or memory, maybe?

**Review by rswarbrick — APPROVED** [review]

**Comment by imphil** [comment]
> Related to this is the behaviour when OTBN reads from the scrambled dmem/imem and doesn't get a grant. This should only happen if a key isn't setup and I believe we'd agreed this would also cause OTBN to halt and signal an error (recoverable rather than fatal alert).

That's a tricky situation to get in actually, as far as I can tell.

We do *not* have memory accesses during a key rotation in the "normal" cases:
- If keys are rotated as part of a user-initiated secure wipe command, the key rotation completes before OTBN returns into idle and allows for software to execute, or external accesses to the memory through the windows are possible.

We can avoid the situation in the startup case:

- If keys are fetched initially during startup, we can delay the start of the software execution until keys are actually present, or alternatively, add a `BUSY_SETUP` STATUS to indicate to Ibex not to start an operation until OTBN is ready.

That leaves one case that I can think of: Key rotation that happens as part of a secure wipe as part of a fatal error handling. Imagine the following scenario:

- Ibex software accesses IMEM through the register window while OTBN is idle.
- This access raises an integrity error, because it detects some ECC mismatch: this is a fatal error.
- The memory window bus access returns an error to Ibex. Ibex somehow ignores the bus error.
- Now secure wipe kicks in and rotates the key.
- A fatal alert is raised.
- Somehow the fatal alert doesn't lead to Ibex being shut down.
- Ibex accesses the OTBN's memory window again.
- Now we have an access to the memory while key rotation is in taking place.

With this scenario being rather constructed (of course there might be others), how about just returning a bus error if the scrambling key isn't valid?

**Comment by tjaychen** [comment]
> > Related to this is the behaviour when OTBN reads from the scrambled dmem/imem and doesn't get a grant. This should only happen if a key isn't setup and I believe we'd agreed this would also cause OTBN to halt and signal an error (recoverable rather than fatal alert).
> 
> That's a tricky situation to get in actually, as far as I can tell.
> 
> We do _not_ have memory accesses during a key rotation in the "normal" cases:
> 
> * If keys are rotated as part of a user-initiated secure wipe command, the key rotation completes before OTBN returns into idle and allows for software to execute, or external accesses to the memory through the windows are possible.
> 
> We can avoid the situation in the startup case:
> 
> * If keys are fetched initially during startup, we can delay the start of the software execution until keys are actually present, or alternatively, add a `BUSY_SETUP` STATUS to indicate to Ibex not to start an operation until OTBN is ready.
> 
> That leaves one case that I can think of: Key rotation that happens as part of a secure wipe as part of a fatal error handling. Imagine the following scenario:
> 
> * Ibex software accesses IMEM through the register window while OTBN is idle.
> * This access raises an integrity error, because it detects some ECC mismatch: this is a fatal error.
> * The memory window bus access returns an error to Ibex. Ibex somehow ignores the bus error.
> * Now secure wipe kicks in and rotates the key.
> * A fatal alert is raised.
> * Somehow the fatal alert doesn't lead to Ibex being shut down.
> * Ibex accesses the OTBN's memory window again.
> * Now we have an access to the memory while key rotation is in taking place.
> 
> With this scenario being rather constructed (of course there might be others), how about just returning a bus error if the scrambling key isn't valid?

i don't remember if @GregAC was specifically checking against a key rotation, or just in general a "memory is not granting my access, this should never happen" error. 

I feel like that specific case is broader than the key rotation case, and "may" also cover unforeseen circumstances. And while normally I would agree with bus_error, it seems like in the other issue (I don't even remember half the things I said) we've already concluded that these types of accesses that should not happen should fall into the alert / status category.  Should we try to stay consistent here? 

I see your point about bus error though, because we don't want the transaction to hang in this case, so we have to make some kind of response.

**Comment by GregAC** [comment]
> i don't remember if @GregAC was specifically checking against a key rotation, or just in general a "memory is not granting my access, this should never happen" error.

Yes that's the concern. From previous discussion here: https://github.com/lowRISC/opentitan/issues/7021 I had thought when rotating keys it'd be possible to start OTBN also, though with what is suggested in this PR: https://github.com/lowRISC/opentitan/pull/7628 that won't be possible.

> If keys are fetched initially during startup, we can delay the start of the software execution until keys are actually present, or alternatively, add a BUSY_SETUP STATUS to indicate to Ibex not to start an operation until OTBN is ready.

I think that would be acceptable. What we were trying to avoid was SW starting OTBN, then OTBN waiting for keys to be available. Keys never appearing so OTBN never finishes and the SW waits forever. Here the start command couldn't be issued so we wouldn't get that scenario. The SW will spin waiting for the status to change but at least it should be more obvious what's causing the hold up.

**Comment by GregAC** [comment]
@imphil can we get the CI issue fixed and this PR merged? I'm working on https://github.com/lowRISC/opentitan/issues/7470 which needs to build on top of it.

**Comment by imphil** [comment]
@GregAC most of the CI issue comes from a lack of implementation :-) But if you're working on #7470 (thanks!) I'll stub out the signals and update the PR this afternoon.

**Comment by GregAC** [comment]
> @GregAC most of the CI issue comes from a lack of implementation :-) 

Ah ok, in that case don't worry. I believe I can add a commit to this PR so we can merge it all together.

File: hw/ip/otbn/data/otbn.hjson (hunk 1 above)
Comment (imphil) [reply hw/ip/otbn/data/otbn.hjson:241]: updated with that wording.

**Comment by imphil** [comment]
@GregAC update here: This change is now complete and passes all CI except for the automated CSR tests in DV. I'll keep looking into that, it's probably some exclusion I missed.

**Comment by imphil** [comment]
Automated CSR tests were failing with a X assert because I forgot to stub out one signal in the err_code register. That's fixed now and we're ready to go to unblock Greg's follow-up doing the actual implementation. I'll need to re-read the discussion here and file follow-ups as necessary; I'll do that next.

**Comment by cdgori** [comment]
Apologies for a drive-by post-merge comment - Tim referenced this issue in another meeting today and I realized I never got around to this one so I went to take a look.

I think the resolution here for this specific / these specific error handling is good, no concerns from my side.

However, what occurred to me in the context of the other meeting was that some types of operations (key generation specifically comes to mind, but there might be others) might want a mechanism to promote recoverable errors to fatal (similar to the -Werror switch in gcc that promotes warnings to errors).  Basically a "paranoid" mode.

Would that be feasible here?

**Comment by GregAC** [comment]
@cdgori an easy enough feature to add, I'll open an issue to discuss.

**Pull request merged**